Mixed Models fixes

FBartos opened the pull request:
fixes: jasp-stats/jasp-test-release#1208
fixes: jasp-stats/jasp-test-release#1207
fixes: jasp-stats/jasp-test-release#1210
fixes: jasp-stats/jasp-test-release#1260
fixes: jasp-stats/jasp-test-release#1326
fixes: jasp-stats/jasp-test-release#1327
fixes: jasp-stats/jasp-test-release#1331
fixes: jasp-stats/jasp-test-release#1331
fixes: jasp-stats/jasp-test-release#1330
fixes: jasp-stats/jasp-test-release#1323

should fix: jasp-stats/jasp-test-release#1329 (the fix works when run from R interactively - the analysis produces the correct error, but it still fails as a development module, can't figure out why though :( )

AlexanderLyNL commented:
I can't install this as a developer module and test :( 

<img width="532" alt="Screenshot 2021-06-09 at 19 13 19" src="https://user-images.githubusercontent.com/10271675/121399507-c8e8c300-c956-11eb-95f9-5913106c7447.png">

AlexanderLyNL commented on R/MixedModelsCommon.R:
```diff
@@ -851,12 +843,7 @@
-            ))), collapse = ":")
-          var_name <-
-            .mmVariableNames(var_name, options$fixedVariables)
+          var_name <- .mmVariableNames(rownames(temp_Corr)[i], options$fixedVariables)
```
tempCorr and varName

Why do we now mix this style into it?

FBartos replied:
it was apparently always there, the only thing that changed was the line break

AlexanderLyNL replied:
You should know my opinion about snake casing by now.

AlexanderLyNL commented on R/MixedModelsCommon.R:
```diff
@@ -1168,6 +1155,11 @@
     }
+
+    if (options$plotsGeom == "geom_violin" && (length(options$plotsAgregatedOver) == 1 && length(unique(dataset[, .v(options$plotsAgregatedOver)])) < 3)) {
+      plots$setError(gettext("Violin geom requires that the random effects grouping factors has at least 3 levels."))
```
The user would not understand what a violin geom is.

FBartos replied:
It's called violin geom in the menu

![image](https://user-images.githubusercontent.com/38475991/121767703-40d30b00-cb5a-11eb-9eec-447c7684fc1f.png)

AlexanderLyNL replied:
That still doesn't make sense to the users.

FBartos replied:
Not sure what you want from me. Change the error message, rename the dropdown, or both? Then we would need a new, different name than ``Background geom`` to describe the geom shape which depicts the background data.

AlexanderLyNL commented on R/MixedModelsCommon.R:
```diff
@@ -2192,12 +2191,18 @@
     
+  if (inherits(model, "error")) {
+    if (model$message == "Dropping columns failed to produce full column rank design matrix")
+      .quitAnalysis(gettext("The specified combination of factors does not produce an estimable model. The most likely reason for this issue is a factor / combination of factors leading to more levels than are estimable."))
```
The specified combination of factors does not produce an estimable model. A factor or combination of factors resulted in more levels than the effective sample size.

FBartos replied:
thanks, changed

AlexanderLyNL reviewed (changes requested):
I'm don't understand why snake cased variable names sneaked into the code.

FBartos commented:
hmm, I'm not getting the jaspGraphs error when using the last nightly

AlexanderLyNL commented:
Perhaps someone else should attempt to review this, since I'm unable to get this to run on my mac.

AlexanderLyNL commented on inst/qml/MixedModelsGLMM.qml:
```diff
@@ -92,12 +92,23 @@ Form {
+			property var familyDefault:
+			{
+				"binomial":			"logit",
+				"binomial_agg":		"logit",
```
snake

FBartos replied:
fixed

AlexanderLyNL reviewed (changes requested):
Basically, the standard style problems appear again. I'll have a better in-depth look later.

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+    return()
+  }
+
+  if (options$plotsGeom == "geom_violin" && (length(options$plotsAgregatedOver) == 1 && length(unique(dataset[, .v(options$plotsAgregatedOver)])) < 3)) {
```
```suggestion
  if (options$plotsGeom == "geom_violin" && (length(options$plotsAgregatedOver) == 1 && length(unique(dataset[, options$plotsAgregatedOver)]) < 3)) {
```
`.v` is no longer needed.

FBartos replied:
removed

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+      "theme_light"   = ggplot2::theme_light()    + ggplot2::theme(legend.position = "bottom"),
+      "theme_minimal" = ggplot2::theme_minimal()  + ggplot2::theme(legend.position = "bottom"),
+      "theme_pubr"    = jaspGraphs::themePubrRaw(legend = options$plotLegendPosition),
+      "theme_apa"     = jaspGraphs::themeApaRaw(legend.pos = switch(
+        options$plotLegendPosition,
+        "none"  = "none",
+        "botom" = "bottommiddle",
+        "right" = "bottomright",
+        "top"   = "topmiddle",
+        "left"  = "bottomleft"
+      ))
```
1. Are you sure that `botom` shouldn't be `bottom`?
2. I find that it leards to nicer code if the values in QML for `$plotLegendPosition` correspond directly to the values that R expects, instead of adding an additional translation step in between. Basically, can't the values for the QML element correspond to `"none"`, `"bottom"`, etc. and then you can delete this switch?

FBartos replied:
1) thanks, it was a typo, fixed

2) it does for all other themes, the APA one does not have those positions

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+    if (options$family == "binomial_agg") {
+      glmmWeight <<- dataset[, .v(options$dependentVariableAggregation)]
+
+      model <- tryCatch(stanova::stanova(
```
shouldn't this and the one in the else statement be `try`?

FBartos replied:
yes, probably all of them, I never get into using try, it's kinda distracting to have the error printed in red but the flow uninterrupted...

vandenman replied:
There's `try(..., silent = TRUE)` if you don't want to see the error message. But `result <- try(...)` is practically identical to `result <- tryCatch(..., error = function(e))`, but much easier to read because there won't be multiple catch statements.

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+        ess_tail   = tempSummary$ess_tail[j]
+      )
+
+      if (varName != "Intercept" && nrow(tempSummary) > 1) {
+        varName <-
+          paste(.unv(unlist(strsplit(
+            as.character(tempSummary$Variable[j]), ","
+          ))), collapse = jaspBase::interactionSymbol)
+        varName <- gsub(" ", "", varName, fixed = TRUE)
+        if (grepl(jaspBase::interactionSymbol, names(modelSummary)[i], fixed = T)) {
+          for (n in unlist(strsplit(.unv(names(
+            modelSummary
+          )[i]), jaspBase::interactionSymbol))) {
+            varName <- gsub(n, "", varName, fixed = TRUE)
+          }
+        } else{
+          varName <-
+            gsub(.unv(names(modelSummary)[i]), "", varName, fixed = TRUE)
+        }
```
This part is a bit odd...

1. you're using `.unv`, but you really shouldn't.
2. a couple lines earlier you're doing `varName <- jaspBase::gsubInteractionSymbol(names(modelSummary)[i])`, so the names of the modelSummary don't contain the interaction symbol. But here you do `grepl(jaspBase::interactionSymbol, names(modelSummary)[i], fixed = T)`? Shouldn't that always be FALSE?
3. Please use `TRUE` instead of `T`.

FBartos replied:
1) fixed
2) the previous lines extract the names and change them, the following lines refer back to the unchanged names (however, there should've been `:` instead of `jaspBase::interactionSymbol`, just fixed that)
The issue is that stanova produces variable names like this: `"facExperim control, facGender f"` and I want to:
a) change the "," to the interaction symbol
b) remove the variable names from the factor levels since I have the factor levels in the title.  
3) sure, fixed

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+    removedMe <- jaspResults[["mmModel"]]$object$removedMe
+    removedTe <- jaspResults[["mmModel"]]$object$removedTe
+    addedRe   <- jaspResults[["mmModel"]]$object$addedRe
+    if (length(removedMe) > 0) {
+      for (j in 1:length(removedMe)) {
+        tempTable$addFootnote(.mmMessageOmmitedTerms1(removedMe[[j]], names(removedMe)[j]),
+                              symbol = gettext("Note:"))
+      }
+    }
```
```suggestion
    for (j in seq_along(removedMe))
      tempTable$addFootnote(.mmMessageOmmitedTerms1(removedMe[[j]], names(removedMe)[j]),
                              symbol = gettext("Note:"))
```
The same goes for the 2 other if statements and loop below this one. In general, `seq_along(x)` is *always* preferred over `1:length(x)`, precisely so you never need this additional if statements.

FBartos replied:
fixed

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+
+  model     <- jaspResults[["mmModel"]]$object$model
+
+  if (options$samplingPlot != "stan_scat") {
+    pars <-
+      paste0(.v(unlist(options$samplingVariable1)), collapse = ":")
+  } else{
+    pars <- c(paste0(.v(unlist(
+      options$samplingVariable1
+    )), collapse = ":"),
+    paste0(.v(unlist(
+      options$samplingVariable2
+    )), collapse = ":"))
+  }
```
`.v` is not needed.

FBartos replied:
removed

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+        paste(.unv(unlist(
+          strsplit(dimnames(samples)$Parameter[cf], ",")
+        )), collapse = ":")
+      coefs_name <- gsub(" ", "", coefs_name, fixed = TRUE)
```
if this is removing leading and/ or trailing whitespace, then `trimws()` is more appropriate (and faster).

FBartos replied:
it's not removing trailing spaces but spaces inside of the characters

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+    for (cf1 in 1:coefs1) {
+      for (cf2 in 1:coefs2) {
+
+        coefs1Name <-
+          paste(.unv(unlist(
+            strsplit(dimnames(samples1)$Parameter[cf1], ",")
+          )), collapse = ":")
```
`.unv`

FBartos replied:
removed

JorisGoosen commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+  }
+  for (v in unlist(options$marginalMeans)) {
+    if (typeof(dataset[, .v(v)]) == "double") {
+      EMMsummary$addColumnInfo(name = .v(v),
```
They are now identity functions os can be removed

FBartos replied:
removed

JorisGoosen commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+  return(plotData)
+
+}
+# as explained in ?is.integer
```
more text is more better

FBartos replied:
`is.integer(1) == FALSE`

FBartos replied:
the documentation of `?is.integer`

> is.integer(x) does not test if x contains integer numbers! For that, use round, as in the function is.wholenumber(x) in the examples.

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+           ifelse(REML, gettext("restricted maximum likelihood"), gettext("maximum likelihood")),
+           ifelse(REML, gettext(" Please note that models with different fixed effects cannot be compared when REML is used. To use ML, switch 'Test model terms' to 'Likelihood ratio tests'."), ""))
+}
+#
+# Copyright (C) 2019 University of Amsterdam
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+
+# TODO: Expose priors specification to users in Bxxx?
+# TODO: Add 3rd level random effects grouping factors ;) (not that difficult actually)
+
+.mmRunAnalysis   <- function(jaspResults, dataset, options, type){
```
what happened here? did you accidentally duplicate the analysis?

FBartos replied:
no, I was just removing the previously present underscores Lexi complained about. (unless you see the same file twice there, but I see only one copy)

FBartos replied:
oh, now I see, I though it was line 1

FBartos replied:
yeah, wtf happened here? thanks for finding it out

vandenman commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+                          "dependentVariableAggregation",
+                          "family",
+                          "link")
+# texts and messages
+.mmMessageInterpretability <-
+  gettext("The intercept corresponds to the (unweighted) grand mean; for each factor with k levels, k - 1 parameters are estimated. Consequently, the estimates cannot be directly mapped to factor levels.")
+.mmMessageSingularFit <-
+  gettext("Model fit is singular. Specified random effects parameters (random intercepts and random slopes) cannot be estimated from the available data. Carefully reduce the random effects structure, but this practice might inflate the reported p-value, and invalidates the analysis.")
```
this is a fine approach, but I'd put these in a separate file.

FBartos replied:
ok, done

vandenman commented on tests/testthat/test-mixedmodelsglmm.R:
```diff
@@ -5,20 +5,20 @@ context("Generalized Linear Mixed Models")
   skip_on_os("linux") # problems with precision outside of windows
-  options <- analysisOptions("MixedModelsGLMM")
-  options$Contrasts <- list(list(isContrast = FALSE, levels = c("1", "2", "3", "4", 
+  options <- jaspTools::analysisOptions("MixedModelsGLMM")
```
just so you know, it's not mandatory to prefix this with `jaspTools::`, although it's not wrong to add it either.

FBartos replied:
I was actually getting errors when running the tests locally without it

vandenman reviewed (changes requested):
Some comments below, I didn't test (yet) if the issues are actually fixed.

Also, concerning the test figures, can you confirm that there are two figures less because these are skipped?

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+           ifelse(REML, gettext("restricted maximum likelihood"), gettext("maximum likelihood")),
+           ifelse(REML, gettext(" Please note that models with different fixed effects cannot be compared when REML is used. To use ML, switch 'Test model terms' to 'Likelihood ratio tests'."), ""))
+}
+#
```
because I accidentally duplicated the file as Don noticed

vandenman commented on R/MixedModelsMessages.R:
```diff
@@ -0,0 +1,209 @@
+    n_bad
+  )
+}
+.mmMessageFitType       <- function(REML) {
+  gettextf("The model was fitted using %1$s.%2$s",
+           ifelse(REML, gettext("restricted maximum likelihood"), gettext("maximum likelihood")),
+           ifelse(REML, gettext(" Please note that models with different fixed effects cannot be compared when REML is used. To use ML, switch 'Test model terms' to 'Likelihood ratio tests'."), ""))
+}
```
if you can avoid `gettextf` it makes the work of the translators a lot easier.
```r
if (REML) {
  gettext("The model was fitted using restricted maximum likelihood. Please note that models with different fixed effects cannot be compared when REML is used. To use ML, switch 'Test model terms' to 'Likelihood ratio tests'.")
} else {
  gettext("The model was fitted using maximum likelihood.")
}
```

FBartos replied:
ok, changed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -498,193 +498,196 @@
+                    df     = afex::nice(model)$df[i])
+
+    if (options$method %in% c("S", "KR")) {
+      tempRow$stat   = model$anova_table$`F`[i]
```
that's how the object is generated by the `afex` package

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -138,34 +138,34 @@
     if(length(options$randomVariables) != 0) options$randomVariables
   ))
   dataset <- dataset[,used_variables]
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -60,28 +60,28 @@
- 
+
     # sampling diagnostics
     if(type %in% c("BLMM", "BGLMM")){
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -20,36 +20,36 @@
+
   # fit the model
   if (.mmReady(options, type)){
     if(type %in% c("LMM", "GLMM")).mmFitModel(jaspResults, dataset, options, type)
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -20,36 +20,36 @@
+
   # fit the model
   if (.mmReady(options, type)){
     if(type %in% c("LMM", "GLMM")).mmFitModel(jaspResults, dataset, options, type)
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -138,34 +138,34 @@
-  
+
   # store the number of missing values into a jaspState object
   n_missing <- createJaspState()
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+
+  trendsVarNames <- colnames(emmTable)[1:(grep(".trend", colnames(emmTable), fixed = TRUE) - 1)]
+
+  for (v in trendsVarNames) {
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+                   probs = c(.5 - options$summaryCI / 2, .5 + options$summaryCI / 2))$summary
+  namesSummary <- rownames(modelSummary)
+  feSummary    <-
+    modelSummary[!grepl("b[", namesSummary, fixed = T) &
```
fixed

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -792,1293 +793,4580 @@
+
+    for (cf in 1:coefs) {
+
+      coefs_name <-
```
fixed

vandenman commented:
@FBartos whenever you're ready just re-request a review!

FBartos commented:
somehow fixes: https://github.com/jasp-stats/jasp-issues/issues/1334

vandenman commented:
@FBartos if you change tryCatch into try then `if (inherits(p, "error"))` should become `if (jaspBase::isTryError(p))` (which is basically `if (inherits(p, "try-error"))`).

FBartos commented:
thanks, and also the error message is no longer under ``$message``

FBartos commented:
finally, should be working now & ready for re-review

FBartos commented:
not sure that this approach was the most efficient - I submitted the PR to fix a set of bugs raised during the testing phase, easily locatable at ~ 150 lines of code, however, ended up doing styling changes unrelated to those fixes resulting in the whole file being basically rewritten and changing things that weren't broken

vandenman commented:
> not sure that this approach was the most efficient

@FBartos yeah the reformatting could have (should have) been done in a separate PR. Now I probably also commented on things I would not have noticed if this PR only consisted of bugfixes which escalated the amount of changes. That being said, you know that JASP uses camelCase and not snake_case, so this was going to end up on your plate sooner or later. It would have also been fine if you had opened a separate issue and/or PR for fixing the style,  (it would also have been been easier for to review), as long as it gets done.

FBartos commented on R/MixedModelsCommon.R:
```diff
@@ -553,27 +553,27 @@
   if (type %in% c("LMM", "GLMM") && !is.null(model)) {
     if (any(attr(model, "class") %in% c("std::runtime_error", "C++Error", "error"))) {
-      if (model$message == "(maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate")
+      if (model == "(maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate")
```
thanks! missed that one

vandenman reviewed (changes requested):
Okay I looked into the issues in JASP:

- [x] https://github.com/jasp-stats/jasp-test-release/issues/1326 is fixed for the frequentist GLM but not for the Bayesian GLM.
- [ ] Bayesian GLMs no longer show output in the model summary table: 
![image](https://user-images.githubusercontent.com/21319932/124596971-2e9b7280-de63-11eb-897a-497b146bf196.png) (https://github.com/jasp-stats/jasp-test-release/issues/1331 seems fixed though) 
- [ ] https://github.com/jasp-stats/jasp-test-release/issues/1323 does not appear fixed, there is now an unexpected error: 
![image](https://user-images.githubusercontent.com/21319932/124597773-2859c600-de64-11eb-9f8b-2dee85638211.png)
- [ ] https://github.com/jasp-stats/jasp-test-release/issues/1330 is not fixed.
- [ ] https://github.com/jasp-stats/jasp-test-release/issues/1331 is not fixed.

vandenman commented:
So if I refresh the file linked in https://github.com/jasp-stats/jasp-test-release/issues/1329#issuecomment-875408115 I get this:

![image](https://user-images.githubusercontent.com/21319932/124779341-fb7be080-df41-11eb-9995-2148c661170b.png)

which is probably not ideal.

I noticed that there you used a lot of internal functions from `rstan`, see e.g., https://github.com/FBartos/jaspMixedModels/blob/05905a922b58802575fdf61f52e12158434c496e/R/MixedModelsCommon.R#L2631-L2636. Well this is the reason why you might not want to do that too often (or at least, write a unit tests for them so you'll notice when they do break).

FBartos commented:
Thanks Don, I will check it out and fix!

FBartos commented:
I can't reproduce this issue: https://github.com/jasp-stats/jasp-test-release/issues/1331

FBartos commented:
ok, I went thru the issues again and they should be working now - sorry, should have done that when I submitted some of the remaining formatting fixes

I could not test it with the latest nightly, it throws a new Rcpp error (https://github.com/jasp-stats/jasp-test-release/issues/1363) , instead, I used nightly from 19:00 07-07-2021 (3eb4d6c0a44b1800f29e71c13735c06ba309500a)

FBartos commented:
The issue within this:
> So if I refresh the file linked in jasp-stats/jasp-test-release#1329 (comment) I get this:

is problematic. It used to drop error directly from the fitting function. Now it's able to finish the fit but the summary function fails to produce the summary. Furtheremore, it's hard to figure out when and why it will create problems - I fitted multiple models and the general tendency was that the models were either wildly misspecified, or the data had extreme scaling. I added a better message suggesting that users check the scaling and factors of their variables, but I don't think that there is much else I can do.

> I noticed that there you used a lot of internal functions from rstan, see e.g., https://github.com/FBartos/jaspMixedModels/blob/05905a922b58802575fdf61f52e12158434c496e/R/MixedModelsCommon.R#L2631-L2636. Well this is the reason why you might not want to do that too often (or at least, write a unit tests for them so you'll notice when they do break).

This particular break wasn't due to rstan internal functions, but the point is still valid. It would be generally good to figure out unit tests for the whole B(G)LMM part of the module. I added an issue to solve that this should be tackled in the future (https://github.com/jasp-stats/jaspMixedModels/issues/48).

Kucharssim commented on inst/qml/MixedModelsBGLMM.qml:
```diff
@@ -81,7 +81,7 @@ Form {
 			[
 				{ label: qsTr("Binomial"),				value: "binomial"},
-				{ label: qsTr("Binomial (aggregated)"),	value: "binomial_agg"},
+				{ label: qsTr("Binomial (aggregated)"),	value: "binomialAgg"},
```
If I am not mistaken, this option was already in the last release. You should add the value name change to `Upgrades.qml` to ensure backward compatibility (see https://github.com/jasp-stats/jasp-desktop/blob/development/Docs/development/jasp-upgrade-qml.md). The same concerns `MixedModelsGLMM.qml` (same option), and the `bootstrap_samples` -> `bootstrapSamples` in `MixedModelsOptions.qml`.

FBartos replied:
thanks, didn't know about this

Kucharssim reviewed (changes requested):
Ufff, so I went through the linked issues, it seems that they are now fixed. 

I went through the code relatively fast, so I did not check every line of code. But it seems generally ok now, including the style changes.

I have one small remark about renaming option values, when you figure that out I think this is ok to be merged.

Kucharssim reviewed (changes requested):
Typically, it's a good idea to test whether your fixes actually fix the issue. It's also good idea to read documentation properly (especially if you do something for the first time). 

Below, suggested fixes to your fixes 😉

FBartos commented:
> Typically, it's a good idea to test whether your fixes actually fix the issue. It's also good idea to read documentation properly (especially if you do something for the first time).
> 
> Below, suggested fixes to your fixes 😉

Thanks, these are all fair points.

To be honest, I did not know how to test it.

FBartos commented:
I tried to propagate the changes to the rest of the update, but the module just fails to be installed in JASP. Any idea why?
@Kucharssim

FBartos commented:
It just crashes JASP hard

Kucharssim commented on inst/Upgrades.qml:
```diff
@@ -0,0 +1,100 @@
+		toVersion:		"0.15"
+
+		ChangeRename { from: "bootstrap_samples";	to: "bootstrapSamples"; }
+		ChangeSetValue
+		{
+			name:		"plotsTheme"
+			condition:	function(options) { return options["plotsTheme"] == "jtools::theme_apa"; }
+			jsonValue:	"theme_apa"
+		}
+		ChangeSetValue
+		{
+			name:		"plotsTheme"
+			condition:	function(options) { return options["plotsTheme"] == "ggpubr::theme_pubr"; }
+			jsonValue:	"theme_pubr"
+		}
```
You could reduce the code duplication by changing the option values in a single function:
```suggestion
		ChangeJS
		{
			name:		"plotsTheme"
			jsFunction:	function(options) 
			{
				switch(options["plotsTheme"])
				{
					case "jtools::theme_apa":	return "theme_apa";
					case "ggpubr::theme_pubr":	return "theme_pubr";
					default:			return options["plotsTheme"]
				}
			}
		}
```

Kucharssim reviewed (commented):
> I did not know how to test it.

Yeah that's understandable. Basically, you need to create a .jasp file with a 0.14.1 JASP version and then load it up after installing your new version of the module. If the options that you changed show something like this
![Screenshot 2021-07-26 at 12 18 47](https://user-images.githubusercontent.com/7093546/126975204-e93a765a-593a-49f3-8b69-285bdfd1acfa.png)
Then the upgrade is not working.

> but the module just fails to be installed in JASP. Any idea why?

No idea, it works for me now just fine... Maybe try Preferences -> Advanced -> Clean installed modules and packages, restart JASP and try install the module again. If that does not work, then I guess it's something for @JorisGoosen to look into.

Otherwise it looks good, I left one minor comment. I guess once we are sure the module is installing correctly, I am fine with merging it.

FBartos commented:
I propagated the changes to the rest, tried removing the installed modules and packages but it didn't help:

![image](https://user-images.githubusercontent.com/38475991/126982295-afdadea9-edac-41ab-9392-8d1ff95f3e6a.png)

Kucharssim commented:
> I propagated the changes to the rest, tried removing the installed modules and packages but it didn't help:

It still does work for me, the upgrades are now also working properly. It could be a bug in a nightly, or something OS specific...? @JorisGoosen @vandenman any ideas?

If this is resolved, I think it's good to be merged.

JorisGoosen commented:
> > I propagated the changes to the rest, tried removing the installed modules and packages but it didn't help:
> 
> It still does work for me, the upgrades are now also working properly. It could be a bug in a nightly, or something OS specific...? @JorisGoosen @vandenman any ideas?

I've been looking into it, some other people also had issues. But for them it seems resolved now. Maybe also for @FBartos?

FBartos commented:
Yes, it actually installed properly on the latest development nightly (I'm not touching that weird Czech stuff :D )

Kucharssim reviewed (approved):
Unless @vandenman or @AlexanderLyNL object, I am willing to merge this monster.

AlexanderLyNL commented:
Unfortunately, I'm still unable to install and test this as a development module, not even with the latest nightly https://github.com/jasp-stats/jasp-desktop/commit/a4e5065e7794bfe783652bf758ccd64c318d644c

JorisGoosen commented:
> Unfortunately, I'm still unable to install and test this as a development module, not even with the latest nightly [jasp-stats/jasp-desktop@a4e5065](https://github.com/jasp-stats/jasp-desktop/commit/a4e5065e7794bfe783652bf758ccd64c318d644c)

https://github.com/jasp-stats/jasp-test-release/issues/1358 then?

JorisGoosen commented:
https://static.jasp-stats.org/Nightlies/MacOS/JASP-nightly-testMixedModelsFixes-e952e31081880b012e00522ddd66d5cb9d6978d6.dmg 

A nightly which includes the translation fixes Frantisek just made, just for you @AlexanderLyNL

FBartos commented:
Can you still install it as a development module? I wanted to fix: https://github.com/jasp-stats/jasp-test-release/issues/1209#issuecomment-895555271 but can't install it even without any additional changes (and I tried 3 different nightlies)

AlexanderLyNL commented:
I still can't install a development module with the new nightly and used Joris's special build to test.

JorisGoosen commented:
Should we maybemerge this and leave additional fixes for later?

This has been going on for long enough now I think. It just gets bigger and bigger

vandenman commented:
sure, let's merge this.
